packages/framework/src/Component/UploadedFile/UploadedFileFacade.php

```diff
@@ -77,18 +77,21 @@ public function manageFiles(object $entity, UploadedFileData $uploadedFileData,
         $uploadedFileTypeConfig = $uploadedFileEntityConfig->getTypeByName($type);
 
         $uploadedFiles = $uploadedFileData->uploadedFiles;
+        $uploadedFileNames = $uploadedFileData->uploadedFileNames;
         $orderedFiles = $uploadedFileData->orderedFiles;
 
+        $this->updateFilesOrder($orderedFiles);
+        $this->updateFileNamesAndSlugs($uploadedFileData->currentFileNamesIndexedById);
+
         if ($uploadedFileTypeConfig->isMultiple()) {
-            $this->updateFilesOrder($orderedFiles);
-            $this->uploadFiles($entity, $uploadedFileEntityConfig->getEntityName(), $type, $uploadedFiles, count($orderedFiles));
+            $this->uploadFiles($entity, $uploadedFileEntityConfig->getEntityName(), $type, $uploadedFiles, count($orderedFiles), $uploadedFileNames);
         } else {
             if (count($orderedFiles) > 1) {
                 array_shift($orderedFiles);
                 $this->deleteFiles($entity, $orderedFiles);
             }
 
-            $this->uploadFile($entity, $uploadedFileEntityConfig->getEntityName(), $type, $uploadedFiles);
+            $this->uploadFile($entity, $uploadedFileEntityConfig->getEntityName(), $type, array_pop($uploadedFiles), array_pop($uploadedFileNames));
```

Comment:
It seems like a way to hell 😕 
It would be much better to pass only `UploadedFileData` (for a single file, not as set of magical arrays of all uploaded files in different properties as is now).
When we or someone else want to use any different values related to files they will need to add another argument and so on. You do not mind that?
I know it is not part of this PR but I need to mention it because it makes me really sad.

Reply:
I partially agree with you, but this method simply requires many data to create files correctly. I am against passing UploadedFileData as you would need then parse it inside of that function.